### ISSUE TITLE
Consider updating for Swift 4.X compatibility

### ISSUE THREAD

**Issue by fwgreen** [DavidSkrundz/Regex#1]
It could take years for the Swift team to update or replace NSRegularExpression.  Until then, there is very little out there and using extension functions to help with NSRegularExpression's clunky API doesn't solve its performance problems with large strings.  Your project is easy to use and doesn't rely on NSRegularExpression or pcre.  Maybe it could be the basis of the new RegularExpression.swift when the time comes?

**Comment by DavidSkrundz** [comment]
I currently don't have plans of updating to swift 4. I have started looking into how I can improve this project, and I've decided to wait for the Swift 5 beta for two main reasons:

1. The current implementation uses NFAs, which gives a quick compilation, but terrible runtime performance when using multiple quantifiers (eg. `(a+b+)+`). I am planning on switching to DFAs which will increase the matching performance, but this will take me some time to write/test.

2. Swift 5 will improve how Strings are used which is very helpful with regular expression processing.

I will re-evaluate this once WWDC begins on June 4th.
